opm/simulators/flow/EclWriter.hpp

```diff
@@ -481,112 +490,149 @@ class EclWriter : public EclGenericWriter<GetPropType<TypeTag, Properties::Grid>
-        std::vector<RestartKey> extraKeys = {{"OPMEXTRA", UnitSystem::measure::identity, false},
-                                             {"THRESHPR", UnitSystem::measure::pressure, inputThpres.active()}};
+            {"SGMAX", UnitSystem::measure::identity, enableNonWettingHysteresis && oilActive && gasActive},
+            {"SHMAX", UnitSystem::measure::identity, enableWettingHysteresis && oilActive && gasActive},
+
+            {"SOMAX", UnitSystem::measure::identity,
+             (enableNonWettingHysteresis && oilActive && waterActive)
+             || simulator_.problem().vapparsActive(simulator_.episodeIndex())},
+
+            {"SOMIN", UnitSystem::measure::identity, enablePCHysteresis && oilActive && gasActive},
+            {"SWHY1", UnitSystem::measure::identity, enablePCHysteresis && oilActive && waterActive},
+            {"SWMAX", UnitSystem::measure::identity, enableWettingHysteresis && oilActive && waterActive},
+
+            {"PPCW", UnitSystem::measure::pressure, enableSwatinit},
+        };
 
         {
             const auto& tracers = simulator_.vanguard().eclState().tracer();
+
             for (const auto& tracer : tracers) {
-                bool enableSolTracer = (tracer.phase == Phase::GAS && FluidSystem::enableDissolvedGas()) ||
-                                       (tracer.phase == Phase::OIL && FluidSystem::enableVaporizedOil());
+                const auto enableSolTracer =
+                    ((tracer.phase == Phase::GAS) && FluidSystem::enableDissolvedGas()) ||
+                    ((tracer.phase == Phase::OIL) && FluidSystem::enableVaporizedOil());
+
                 solutionKeys.emplace_back(tracer.fname(), UnitSystem::measure::identity, true);
                 solutionKeys.emplace_back(tracer.sname(), UnitSystem::measure::identity, enableSolTracer);
             }
         }
 
-        // The episodeIndex is rewined one back before beginRestart is called
-        // and can not be used here.
-        // We just ask the initconfig directly to be sure that we use the correct
-        // index.
-        const auto& initconfig = simulator_.vanguard().eclState().getInitConfig();
-        int restartStepIdx = initconfig.getRestartStep();
+        const auto& inputThpres = eclState().getSimulationConfig().getThresholdPressure();
+        const std::vector<RestartKey> extraKeys {
+            {"OPMEXTRA", UnitSystem::measure::identity, false},
+            {"THRESHPR", UnitSystem::measure::pressure, inputThpres.active()},
+        };
 
-        const auto& gridView = simulator_.vanguard().gridView();
-        unsigned numElements = gridView.size(/*codim=*/0);
-        outputModule_->allocBuffers(numElements, restartStepIdx, /*isSubStep=*/false, /*log=*/false, /*isRestart*/ true);
+        const auto& gridView = this->simulator_.vanguard().gridView();
+        const auto numElements = gridView.size(/*codim=*/0);
+
+        {
+            // The episodeIndex is rewined one back before beginRestart is
```

Comment:
rewound

Reply:
> rewound

Good catch.  I've pushed an update to fix that.